### ISSUE TITLE
Partial type fails to narrow in spite of check for undefined II

### ISSUE THREAD

**Issue by NoelAbrahams** [microsoft/TypeScript#31495]
**TypeScript Version:**  3.4.0
**Search Terms:** Partial switch 

**Code**
```ts
const enum BarEnum {
    bar1 = 1,
    bar2 = 2,
}

type TypeBar1 = {

    foo: BarEnum.bar1;
    array: number[];
};

type TypeBar2 = {

    foo: BarEnum.bar2;
    array: number[];
};

type UnionOfBar = TypeBar1 | TypeBar2;

const value: Partial<UnionOfBar> = {} as any;

if (value.array) {

    value.array.toString(); // Okay

    switch (value.foo) {

        case BarEnum.bar1:
            value.array.toString(); // Error: Object is possibly undefined
            break;
    }
}
```

**Expected behavior:**
No compiler error

**Actual behavior:**
The compiler complains of `value.array` being undefined

**Playground Link:** 
Not able to reproduce on the playground. I think this is a bug in 3.4.

**Related Issues:** 
This is another version of the bug reported here: #29496

**Comment by j-oliveras** [comment]
It can be reproduced on [playground](https://www.typescriptlang.org/play/#src=const%20enum%20BarEnum%20%7B%0D%0A%20%20%20%20bar1%20%3D%201%2C%0D%0A%20%20%20%20bar2%20%3D%202%2C%0D%0A%7D%0D%0A%0D%0Atype%20TypeBar1%20%3D%20%7B%0D%0A%0D%0A%20%20%20%20foo%3A%20BarEnum.bar1%3B%0D%0A%20%20%20%20array%3A%20number%5B%5D%3B%0D%0A%7D%3B%0D%0A%0D%0Atype%20TypeBar2%20%3D%20%7B%0D%0A%0D%0A%20%20%20%20foo%3A%20BarEnum.bar2%3B%0D%0A%20%20%20%20array%3A%20number%5B%5D%3B%0D%0A%7D%3B%0D%0A%0D%0Atype%20UnionOfBar%20%3D%20TypeBar1%20%7C%20TypeBar2%3B%0D%0A%0D%0Aconst%20value%3A%20Partial%3CUnionOfBar%3E%20%3D%20%7B%7D%20as%20any%3B%0D%0A%0D%0Aif%20(value.array)%20%7B%0D%0A%0D%0A%20%20%20%20value.array.toString()%3B%20%2F%2F%20Okay%0D%0A%0D%0A%20%20%20%20switch%20(value.foo)%20%7B%0D%0A%0D%0A%20%20%20%20%20%20%20%20case%20BarEnum.bar1%3A%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20value.array.toString()%3B%20%2F%2F%20Error%3A%20Object%20is%20possibly%20undefined%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20break%3B%0D%0A%20%20%20%20%7D%0D%0A%7D) enabling `strictNullChecks` under Options.

Probably related to #9998.

**Comment by jack-williams** [comment]
Narrowing by discriminant (which is happening at `value.foo` in the switch) narrows from the declared type of `value`, thus ignoring the previous property narrowing obtained from `value.array`. This is probably a design limitation / working as intended.

**Comment by NoelAbrahams** [comment]
It may be a design limitation, but it certainly cannot be working as intended - unless the intention is to surprise the user!

**Comment by design2soft** [comment]
I am facing the same issue.

**Comment by sigorilla** [comment]
+1, [example](https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgJICE4GcUG8BQyRyMwEANgCZYD8AXMgcc8pAB5gNZhSgDmAbkLEAvkJH58oSLEQpUAEThg4AQWQQOEENTSYcjYUTABPAA4QGAcgD2ICFfGTp0eEjRKV6DVp1Y92HhGrOaWyFZgAO42jvgS+DAAriAIYMB2yGZwUDiecAAUlMpwDIrF6gA+HsXoAJRcPPyGzMAwyPkAhEUqAHSkFNS1zSxEUBBgiVAg4bHM8cxjE1PI3XA9phbIALw74XYOyDQrxX1kVFjrmmDI1rESQA)

**Comment by threehams** [comment]
The examples in this issue are fixed in 3.8. I don't know what change fixed them, though.

**Comment by NoelAbrahams** [comment]
@threehams, you're right: it's been fixed. Thanks.
